### PR TITLE
Adding tests for ZF3 Zend\EventManager

### PR DESCRIPTION
**Pull request by asgrim**
**Stuff Todo**
- [x] `\Zend\EventManager\Exception\RuntimeException`
- [x] `\Zend\EventManager\Factory\ListenerPluginManagerFactory` (ignore)
- [x] `\Zend\EventManager\Event`
- [x] `\Zend\EventManager\EventManager`
- [x] `\Zend\EventManager\ResponseCollection`

### CODE REVIEW

**Comment by asgrim** [comment]
@bakura10 I have a query on how to be lazy...

``` php
<?php

// ... blah

class EventManagerTest extends \PHPUnit_Framework_TestCase
{
    // ... snip ...

    public function testLazyInstantiator()
    {
        $hasBeenTriggered = false;

        $myEventListener = function (Event $event) {
            $hasBeenTriggered = true;
        };

        $eventManager = new EventManager(function ($requestedListener) use ($myEventListener) {
            $this->assertSame('MyEventListener', $requestedListener);
            return $myEventListener;
        });

        $eventManager->attach('MyEvent', /* what is this supposed to be to be lazy??? */);

        // I tried:
        //  - ['classname', 'function']
        //  - [this, 'function']
        //  - ['something', function (..) {.. blah ..}]
        // etc. with various errors, e.g. when trigger() is called it complains about $listener...

        $event = new Event();
        $eventManager->trigger('MyEvent', $event);

        $this->assertTrue($hasBeenTriggered);
    }
}
```

**Comment by asgrim** [comment]
https://gist.github.com/asgrim/60c73a0fffab11c60528

**Comment by codeliner** [comment]
@asgrim I guess ['classname', 'function'] is the desired way. What error do you get when trying this?

**Comment by asgrim** [comment]
Yes - sorted now, I chatted with @bakura10 last night - I'm working on it later, but basically if it's `['classname', 'function']` it gets classed as lazy, so the lazy instantiation kicks in... if it is `[$alreadyAnInstance, 'function']` it does not need to. :D (see the gist in my last comment btw for a working test)

**Comment by codeliner** [comment]
:+1: ok cool

**Comment by asgrim** [comment]
ping @bakura10 - ready for review/merge/feedback :tada:

**Comment by asgrim** [comment]
I've rebased to your `refactor` branch also

File: test/EventManagerTest.php [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            'listenerThree',
+            'listenerTwo',
+        ], $listenersTriggered);
+    }
+
+    public function testCreatingEventManagerWithoutListenerInstantiatorCausesExceptionWhenEventsFired()
+    {
+        $eventManager = new EventManager();
+
+        $spec = [
```
Comment (bakura10) [inline test/EventManagerTest.php]: Can be removed, the exception is thrown in the constructor of the EVM

File: test/EventManagerTest.php [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            'SomeEventListener'
+        ];
+        $eventManager->attach('MyEvent', $spec);
+
+        $this->setExpectedException(RuntimeException::class);
+
+        $eventManager->trigger('MyEvent', new Event());
```
Comment (bakura10) [inline test/EventManagerTest.php:58]: Can be removed too.

File: test/EventManagerTest.php (hunk 1 above)
Comment (asgrim) [inline test/EventManagerTest.php]: No, it's thrown in a closure which is assigned in the constructor... the `$this->listenerInstantiator` isn't actually called (thus causing the exception) until when `trigger` is called

Comment (asgrim) [inline test/EventManagerTest.php]: Specifically, it must `attach` a lazy listener and then `trigger` the even to call the `listenerInstantiator` (which was not provided in the constructor, so will cause the exception)

File: test/EventManagerTest.php (hunk 2 above)
Comment (asgrim) [inline test/EventManagerTest.php:58]: See above comments ^_^

**Comment by bakura10** [comment]
Looks nice!

Some tests are missing:
- Test wildcard listeners
- Make sure that wild card listeners of same priority are always triggered AFTER the named event (so if one listener is attached to event "my.event" with priority 1 and one listener is attached to event "*" with priority 1, listener attached to "my.event" must be exeucted before, even if it was attached after the wildcard.

**Comment by asgrim** [comment]
yes I can add these (possibly not tonight as going to a UG meetup, depends how late I get back)

**Comment by bakura10** [comment]
I'm going to merge that, you'll can add the missing tests in another PR :D.

**Comment by asgrim** [comment]
np

**Pull request merged**